transports/quic/src/transport.rs

```diff
@@ -198,29 +205,42 @@ impl<P: Provider> GenTransport<P> {
 }
 
 impl<P: Provider> Transport for GenTransport<P> {
-    type Output = (PeerId, Connection);
+    type Output = (PeerId, StreamMuxerBox);
     type Error = Error;
-    type ListenerUpgrade = Connecting;
+    type ListenerUpgrade = BoxFuture<'static, Result<Self::Output, Self::Error>>;
     type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
 
     fn listen_on(
         &mut self,
         listener_id: ListenerId,
         addr: Multiaddr,
     ) -> Result<(), TransportError<Self::Error>> {
-        let (socket_addr, version, _peer_id) = self.remote_multiaddr_to_socketaddr(addr, false)?;
-        let endpoint_config = self.quinn_config.endpoint_config.clone();
-        let server_config = self.quinn_config.server_config.clone();
-        let socket = self.create_socket(socket_addr).map_err(Self::Error::from)?;
+        let (socket_addr, version, _peer_id, is_webtransport_addr) = self.remote_multiaddr_to_socketaddr(addr, false)?;
 
+        let socket = self.create_socket(socket_addr).map_err(Self::Error::from)?;
         let socket_c = socket.try_clone().map_err(Self::Error::from)?;
-        let endpoint = Self::new_endpoint(endpoint_config, Some(server_config), socket)?;
+
+        let (server_config, noise, cert_hashes) = self.config.server_quinn_config()
+            .map_err(|e| Error::CertificateGenerationError(e))?;
+
+        let endpoint = match self.get_existed_endpoint(&socket_addr, &cert_hashes) {
+            Some(res) => res,
+            None => Self::new_endpoint(self.config.endpoint_config(), Some(server_config), socket)?,
+        };
```

Comment:
I am not sure this will work as intended. Using the same `endpoint` in two `Listener`s will make them compete, no?

Let's consider the following scenario:

1. User calls `listen_on` for `/quic-v1`
2. User calls `listen_on` for `/webtransport` using the same port
3. Another peer attempts to connect using the `/webtransport` address

The incoming connection will be picked up by one of the two listeners. Until we inspect the `Connecting` struct, we don't actually know whether it is a QUIC or a WebTransport connection.

So I think what we need to do is:

- At most have 1 `Listener` for a given socket address
- The listener needs to be one of 3 modes:
	- Only QUIC
	- Only WT
	- Both
- When we receive a `listen_on` for a socket with an existing listener, we need to _update_ its mode
- For incoming connections, we need to inspect it whether it is a QUIC or WT connection and check whether the listener supports whatever is incoming. If yes, upgrade it accordingly.

What do you think?